### PR TITLE
update to d3@7.4

### PR DESCRIPTION
**Pull request by Fil**
Updates:

- [x] d3-array to v3.1.6
- [x] d3-shape to v3.1.0
- [x] d3-hierarchy to v3.1.2
- [x] d3-color to v3.1.0

### d3-array

https://github.com/d3/d3-array/releases/tag/v3.1.6

d3.rank can now take a comparator in addition to an accessor
d3.rank, d3.sort, d3.bisector, and d3.groupSort now require comparators to have exactly two arguments.

### d3-shape

https://github.com/d3/d3-shape/releases/tag/v3.1.0

Add d3.symbolsStroke.
Add d3.symbolsFill, replacing (deprecating) d3.symbols.
Add d3.symbolAsterisk.
Add d3.symbolDiamond2.
Add d3.symbolPlus.
Add d3.symbolSquare2.
Add d3.symbolTriangle2.
Add d3.symbolX.
Add d3.link.

### d3-hierarchy

https://github.com/d3/d3-hierarchy/releases/tag/v3.1.2

Add stratify.path.

### d3-color

https://github.com/d3/d3-color/releases/tag/v3.1.0

Add [rgb.clamp](https://github.com/d3/d3-color/blob/main/README.md#rgb_clamp) and [hsl.clamp](https://github.com/d3/d3-color/blob/main/README.md#hsl_clamp). https://github.com/d3/d3-color/pull/102
Add [color.formatHex8](https://github.com/d3/d3-color/blob/main/README.md#color_formatHex8). https://github.com/d3/d3-color/pull/103

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`

If changing an existing definition:
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by Methuselah96** [comment]
Hey @Fil, glad to see you over here. The pattern I've tried to follow in the past is to one minor version of d3 per PR. That way we don't miss adding anything from previous versions and the d3 version number is as close to accurate as possible. For example, this PR adds stuff from d3 7.3 even though the published version of d3 will still only be 7.1 if this is merged.

As an example of how I've done it in the past here is: [6.4](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52902), [6.5](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/53092), [6.6](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/53156), [6.7](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/53178), and [7.1](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56119).

If you're interested in maintaining the types consistently like that, then that would be awesome. Otherwise, I can try to pick back up the work and get the latest few versions done.

**Comment by Fil** [comment]
Meaning, if I want to update d3-shape to @3.1 (as I did here), I also need to upgrade the types for all the other modules that have had changes between d3@7.1 (last release that was "typed") and d3@7.4 (current version of D3)?

If this is correct, and double-checking https://github.com/d3/d3/releases, this means I need to update:
- **d3-color**: Add [rgb.clamp](https://github.com/d3/d3-color/blob/main/README.md#rgb_clamp) and [hsl.clamp](https://github.com/d3/d3-color/blob/main/README.md#hsl_clamp); add [color.formatHex8](https://github.com/d3/d3-color/blob/main/README.md#color_formatHex8).
- **d3-hierarchy**: Add [stratify.path](https://github.com/d3/d3-hierarchy/blob/main/README.md#stratify_path).
- **d3-array**: d3.rank can now take a comparator in addition to an accessor; d3.rank, d3.sort, d3.bisector, and d3.groupSort now require comparators to have exactly two arguments.

right?

**Comment by Methuselah96** [comment]
Right, although you don't need to go to 7.4, since these d3-shape changes were made in 7.3. So the PRs I would imagine would be in this order:

- A PR for changes included in d3 7.1.1 (the d3-array changes you mentioned)
- A PR for changes included in d3 7.2 (stratify.path)
- A PR for changes included in d3 7.3 (basically this PR)
- (optionally) A PR for changes included in d3 7.4 (d3-color changes)

I'm totally willing to do the other PRs myself, this is just the easiest way for us to make sure we aren't missing any type changes. Let me know if you're up for doing the earlier versions or if you would like me to.

**Comment by Fil** [comment]
I'm working on d3-array now, but I'm very slow, so in the meantime feel free to do d3-hierarchy and d3-color?

**Comment by Fil** [comment]
d3-array done!

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## d3-array ([<kbd>unpkg</kbd>](https://unpkg.com/browse/d3-array@latest/))
was missing the following properties:
1. scan
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 93dd41bbf1dcb4c48fd4589365381f8351a94e86
</p>

**Comment by Fil** [comment]
what is this missing "scan" property? 👀

**Comment by peterblazejewicz** [comment]
> * I don't understand what @DangerBotOSS is complaining about in the comment above

It analyzes content of the module, it can show false positives at some scanerios. In short, it detects exported symbols in a module and compares with exported one from types. You can skip this one as a niptick, thx!

File: types/d3-color/index.d.ts [hunk 1]
```diff
@@ -6,9 +6,10 @@
 //                 denisname <https://github.com/denisname>
 //                 Hugues Stefanski <https://github.com/ledragon>
 //                 Nathan Bierema <https://github.com/Methuselah96>
+//                 Fil <https://github.com/Fil>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
-// Last module patch version validated against: 3.0.1
+// Last module patch version validated against: 3.1.0
```
Comment (Methuselah96) [inline types/d3-color/index.d.ts:12]: You need to bump the version number on line 1 of this file as well.

File: types/d3-hierarchy/index.d.ts [hunk 2]
```diff
@@ -5,9 +5,10 @@
 //                 Boris Yankov <https://github.com/borisyankov>
 //                 denisname <https://github.com/denisname>
 //                 Nathan Bierema <https://github.com/Methuselah96>
+//                 Fil <https://github.com/Fil>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
-// Last module patch version validated against: 3.0.1
+// Last module patch version validated against: 3.1.2
```
Comment (Methuselah96) [inline types/d3-hierarchy/index.d.ts:11]: Same here.

**Review by Methuselah96 — CHANGES_REQUESTED** [review]
Can you also bump the version numbers in `d3/index.d.ts`?

File: types/d3-hierarchy/index.d.ts (hunk 2 above)
Comment (Fil) [reply types/d3-hierarchy/index.d.ts:11]: done, thank you!

File: types/d3-color/index.d.ts (hunk 1 above)
Comment (Fil) [reply types/d3-color/index.d.ts:12]: done

**Review by Methuselah96 — APPROVED** [review]
Awesome, thanks for your work on this!

**Review by peterblazejewicz — APPROVED** [review]
@Fil thanks!

**Comment by Fil** [comment]
Ready to merge

**Pull request merged**